### PR TITLE
SPINEDEM-4939 Handle custom attributes header

### PR DESCRIPTION
**Pull request by fabianmcgibbon**
## Summary
* Routine Change
* :exclamation: Breaking Change
* :robot: Operational or Infrastructure Change
* :sparkles: New Feature
* :warning: Potential issues that might be caused by this change

Add any other relevant notes or explanations here. **Remove this line if you have nothing to add.**

## Reviews Required
* [x] Dev
* [ ] Test
* [ ] Tech Author
* [ ] Product Owner

## Review Checklist
:information_source: This section is to be filled in by the **reviewer**.

* [ ] I have reviewed the changes in this PR and they fill all or part of the acceptance criteria of the ticket, and the code is in a mergeable state.
* [ ] If there were infrastructure, operational, or build changes, I have made sure there is sufficient evidence that the changes will work.

### CODE REVIEW

File: proxies/live/apiproxy/resources/jsc/ExtractCustomAttributes.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+function extractCustomAttributes(json) {
+    if (json) {
+        try {
+            var parsed = JSON.parse(json);
+            if (parsed.pds && parsed.pds["custom-attributes"]) {
+                context.setVariable("extractedCustomAttributes", JSON.stringify(parsed.pds["custom-attributes"]))
+            }
+        } catch (e) {
+            console.log("exception thrown extracting custom attributes: ", e)
+        }
```
Comment (davidhamill1-nhs) [inline proxies/live/apiproxy/resources/jsc/ExtractCustomAttributes.js:10]: Not sure how we should handle exceptions in the proxy layer.

I guess the only way an error should occur here is by someone not setting the object correctly in Apigee. If this does happen, it feels like we should continue raising the error so that it can be seen and handled. 

Might need to find an example of how APIM do this in other JS policies, if at all.

File: proxies/live/apiproxy/targets/ig3.xml [hunk 2]
```diff
@@ -78,6 +81,12 @@
       <Step>
         <Name>AssignMessage.AddAsidHeader</Name>
       </Step>
+      <Step>
+        <Name>AssignMessage.RemoveCustomAttributesHeader</Name>
+      </Step>
```
Comment (davidhamill1-nhs) [inline proxies/live/apiproxy/targets/ig3.xml:86]: Do we need to apply this on the way back out?

File: proxies/live/apiproxy/resources/jsc/ExtractCustomAttributes.js (hunk 1 above)
Comment (fabianmcgibbon) [reply proxies/live/apiproxy/resources/jsc/ExtractCustomAttributes.js:10]: Don't see this being done anywhere else, do you think it would be more useful to set a extractedCustomAttributesException variable with the exception as the value?

File: proxies/live/apiproxy/targets/ig3.xml (hunk 2 above)
Comment (fabianmcgibbon) [reply proxies/live/apiproxy/targets/ig3.xml:86]: I don't believe so, the header isn't present on the response by default

Comment (davidhamill1-nhs) [reply proxies/live/apiproxy/targets/ig3.xml:86]: I guess Spine proactively writes back the headers, rather than just mirroring what has been sent 👍

Comment (davidhamill1-nhs) [reply proxies/live/apiproxy/targets/ig3.xml:86]: @agelledi Would be good to have this asserted in a test.

File: proxies/live/apiproxy/resources/jsc/ExtractCustomAttributes.js (hunk 1 above)
Comment (davidhamill1-nhs) [reply proxies/live/apiproxy/resources/jsc/ExtractCustomAttributes.js]: I suspect the error may still get obfuscated by the default error created by the proxy...

Comment (davidhamill1-nhs) [reply proxies/live/apiproxy/resources/jsc/ExtractCustomAttributes.js]: https://github.com/NHSDigital/personal-demographics-service-api/blob/7920188d535fb4e7c003d8e7fe9f42ab3225123c/proxies/live/apiproxy/policies/AssignMessage.Errors.Default.xml

Comment (davidhamill1-nhs) [reply proxies/live/apiproxy/resources/jsc/ExtractCustomAttributes.js]: We don't want to pass back to the consumer the internal error message. So the generic UNKNOWN_ERROR would be good for their response. It is more for the investigation, is there a way we can get that logged?

Comment (fabianmcgibbon) [reply proxies/live/apiproxy/resources/jsc/ExtractCustomAttributes.js]: Ah correct, I will remove from the try catch

File: karate-tests/src/test/java/helpers/utils.feature [hunk 3]
```diff
@@ -273,4 +273,40 @@ Scenario:
...
+        for (var j = 0; j < lines.length; j++) {
+          if (lines[j] === null) {
+            hasNull = true
+            if (!_lines[j] || !_lines[j].extension || !_lines[j].extension.length) {
+              return false
+            }
+          }
+        }
+      }
+      return hasNull
```
Comment (fabianmcgibbon) [inline karate-tests/src/test/java/helpers/utils.feature:299]: nit: Is this equivalent? It just seems a bit easier to understand to me

```
function(addresses) {
      for (var i = 0; i < addresses.length; i++) {
        var lines = addresses[i].line
        var _lines = addresses[i]._line
        for (var j = 0; j < lines.length; j++) {
          if (lines[j] === null) {
            if (!_lines[j] || !_lines[j].extension || !_lines[j].extension.length) {
              return false
            }
          }
        }
      }
      return true
```

Comment (agelledi) [reply karate-tests/src/test/java/helpers/utils.feature:299]: This function returns` false `if the address does not contain any null entries. It returns `true` only when the address includes null entries, and each null entry is accompanied by a corresponding _line extension

File: proxies/live/apiproxy/targets/ig3.xml (hunk 2 above)
Comment (agelledi) [reply proxies/live/apiproxy/targets/ig3.xml:86]: added header assertion `assert utils.validateResponseHeaders(requestHeaders, responseHeaders)`

File: karate-tests/src/test/java/patients/emptyAddressLines/updatePatientDetails.feature
Comment (davidhamill1-nhs) [inline karate-tests/src/test/java/patients/emptyAddressLines/updatePatientDetails.feature:1]: Maybe one for future work rather than for this. Given this is using healthcare worker access mode, would it make test data cleaner to create a record and then update that newly created record? Rather than adding then removing temp addresses to the same record.

File: proxies/live/apiproxy/targets/ig3.xml (hunk 2 above)
Comment (davidhamill1-nhs) [reply proxies/live/apiproxy/targets/ig3.xml:86]: 1. `utils.validateResponseHeaders()` takes only one argument, the request headers, and pulls the response headers from the karate context. Does it do anything with the second argument?
2. `utils.validateResponseHeaders()` only validates that the response request and correlation ids match the request's. It will not check that NHSE-PDS-Custom-Attributes is not present.

File: karate-tests/src/test/java/patients/emptyAddressLines/updatePatientDetails.feature [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+   * match utils.checkNullsHaveExtensions(addresses) == true
+ 
+Scenario: Updating contact details response should show empty address lines
+    * def nhsNumber = '9733162264'
+    * def requestHeaders = call read('classpath:auth/auth-headers.js')
+    * configure headers = requestHeaders
+    * path 'Patient', nhsNumber
+    * method get
+    * status 200
+    * def originalVersion = parseInt(response.meta.versionId)
```
Comment (davidhamill1-nhs) [inline karate-tests/src/test/java/patients/emptyAddressLines/updatePatientDetails.feature:89]: This doesn't appear to be used.
```suggestion
```

File: proxies/live/apiproxy/targets/ig3.xml (hunk 2 above)
Comment (davidhamill1-nhs) [reply proxies/live/apiproxy/targets/ig3.xml:86]: 3. Could we write a test that ensures an app without the attribute sends a request with the header applied themselves, they *do not* get the null lines returned.

Comment (agelledi) [reply proxies/live/apiproxy/targets/ig3.xml:86]: Thanks for highlighting this , I have added a test - app without the attribute sends a request with the header applied themselves, they do not get the null lines returned and I have included the header validation check 
` * match responseHeaders['Nhse-Pds-Custom-Attributes'] == '#notpresent'`
`Add empty address lines custom attribute for default test app - response should not include null address lines `

File: karate-tests/src/test/java/patients/emptyAddressLines/updatePatientDetails.feature (hunk 4 above)
Comment (agelledi) [reply karate-tests/src/test/java/patients/emptyAddressLines/updatePatientDetails.feature]: Removed unused variable

File: karate-tests/src/test/java/patients/emptyAddressLines/updatePatientDetails.feature [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    * def tempAddressDetails = utils.removeNullsFromAddress(response.address.find(x => x.use == "temp"))
+    * def addresses = response.address
+    * match utils.checkNullsHaveExtensions(addresses) == true
+
+    
+   # Test fails if the patient's temp address details are not present in the record
+  
+   * if (tempAddressDetails == null) {karate.fail('No value found for temporary address, stopping the test.')}
+   * def addressIndex = response.address.findIndex(x => x.use == "temp")
+   * def tempAddressPath =  "/address/" + addressIndex
```
[the remark this answers is not in the document]
Comment (agelledi) [reply karate-tests/src/test/java/patients/emptyAddressLines/updatePatientDetails.feature]: Added suggested fix

File: karate-tests/src/test/java/patients/emptyAddressLines/updatePatientDetails.feature [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    * def utils = call read('classpath:helpers/utils.feature')
+    * def faker = Java.type('helpers.FakerWrapper')       
+Scenario: Updating temporary address response should show empty address lines
+    * def nhsNumber = '9733162256'
+    * path 'Patient', nhsNumber
+    * method get
+    * status 200
+    * def originalVersion = parseInt(response.meta.versionId)
+    * def originalEtag = karate.response.header('etag')
+    * def istempAddressOnRecord = response.address.find(x => x.use == "temp")
```
[the remark this answers is not in the document]
Comment (agelledi) [reply karate-tests/src/test/java/patients/emptyAddressLines/updatePatientDetails.feature]: Removed `originalVersion` and` istempAddressOnRecord`

File: karate-tests/src/test/java/patients/emptyAddressLines/postPatient.feature [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+  * def faker = Java.type('helpers.FakerWrapper')
+  * json Period = karate.readAsString('classpath:schemas/Period.json')
+  * json addressSchema = karate.readAsString('classpath:schemas/Address.json') 
+    
+  * def accessToken = karate.call('classpath:auth/auth-redirect.feature', {clientID: karate.get('emptyAddressLinesClientID'), clientSecret:karate.get('emptyAddressLinesClientSecret')}).accessToken
+  * def requestHeaders = call read('classpath:auth/auth-headers.js')
+  * configure headers = requestHeaders  
+  * url baseURL
+
+  # Use this family name if the test is going to create patients - this is used by a batch job that cleans up the system database
```
[the remark this answers is not in the document]
Comment (agelledi) [reply karate-tests/src/test/java/patients/emptyAddressLines/postPatient.feature]: Added suggested change

File: karate-tests/src/test/java/helpers/utils.feature (hunk 3 above)
Comment (agelledi) [reply karate-tests/src/test/java/helpers/utils.feature:305]: Included non-null address lines have a corresponding null element in the _line array check

**Review by davidhamill1-nhs — APPROVED** [review]
